### PR TITLE
Adding cloud tests

### PR DESCRIPTION
**Pull request by mzitnik**
## Summary
Adding cloud tests coverage

## Checklist
Delete items not relevant to your PR:
- [ ] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

### CODE REVIEW

File: clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java [hunk 1]
```diff
@@ -134,7 +137,8 @@ public void testIssue1373() throws SQLException {
         String sql = String.format("INSERT INTO %s (%s) SELECT %s FROM input %s", TABLE_NAME, columnNames, columnNames, columnValues);
         Connection conn = dataSource.getConnection("default", "");
         Statement st = conn.createStatement();
-        st.execute(String.format("CREATE TABLE %s (`event_id` String, `num01` Int8, `event_id_01` String) ENGINE = Log", TABLE_NAME));
+        st.execute(String.format("DROP TABLE IF EXISTS %s", TABLE_NAME));
```
Comment (chernser) [inline clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java:140]: I think it will become a problem for concurrent builds because tests will use shared DB and table names for each test run are not unique. 
Adding timestamp to the table might solve the issue, but it requires extra steps for tables cleanup (if tests were aborted)

Comment (mzitnik) [reply clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java:140]: Agree, but as you can see it still in draft

Comment (chernser) [reply clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java:140]: Right. I've commented just to have a discussion about it.

Comment (mshustov) [reply clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java:140]: js client uses the same technic as @chernser proposed https://github.com/ClickHouse/clickhouse-js/blob/abb35e9c76377a2a688bb2e4705df13749143b71/packages/client-common/__tests__/utils/client.ts#L89

go client just creates a new instance using terraform https://github.com/ClickHouse/clickhouse-go/blob/b576bdef50952f0f6b8a26e5e6f641ca549b71f1/.github/workflows/run-tests-cloud.yml

**Comment by chernser** [comment]
I think, this change would be good to adjust for running tests against any host not only cloud.
Current version assumes that cloud using https protocol but test host may be another cases. At least correcting this thing would be a great addition. 
As for adding to CI - I think we create a separate workflow. Then we may merge it back. We may not require to run tests against cloud, I think.

File: clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseServerForTest.java [hunk 2]
```diff
@@ -289,5 +379,44 @@ public static void afterSuite() {
         if (clickhouseContainer != null) {
             clickhouseContainer.stop();
         }
+
+        if (isCloud()) {
+            ClickHouseNode server = ClickHouseNode.builder(ClickHouseNode.builder().addOption(ClickHouseClientOption.SSL.getKey(), "true").build())
```
Comment (mshustov) [inline clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseServerForTest.java:384]: That looks like a lot of logic, should we consider moving it put of the existing class?

File: clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseServerForTest.java [hunk 3]
```diff
@@ -289,5 +379,44 @@ public static void afterSuite() {
...
+                    .address(ClickHouseProtocol.HTTP, new InetSocketAddress(System.getenv("CLICKHOUSE_CLOUD_HOST"), 8443))
+                    .credentials(new ClickHouseCredentials("default", getPassword()))
+                    .options(Collections.singletonMap(ClickHouseClientOption.SSL.getKey(), "true"))
+                    .build();
+
+            //Remove database after testing
+            boolean success = false;
+            int retries = 0;
```
Comment (mshustov) [inline clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseServerForTest.java:392]: Could we extract retry logic outside? reading inlined code that mixes responsibilities is hard

File: clickhouse-client/src/test/java/com/clickhouse/client/ClientIntegrationTest.java [hunk 4]
```diff
@@ -2599,6 +2674,7 @@ public void testLoadBalancingPolicyFailover(ClickHouseLoadBalancingPolicy loadBa
 
     @Test(groups = {"integration"})
     public void testFailover() throws ClickHouseException {
+        if (isCloud()) return; //TODO: testFailover - Revisit
```
Comment (mshustov) [inline clickhouse-client/src/test/java/com/clickhouse/client/ClientIntegrationTest.java:2677]: Can we add a link to https://github.com/ClickHouse/clickhouse-java/issues/1747 ? It would help us to search for these disabled tests. Even though, I don't think we're going to fix all of them

File: clickhouse-http-client/src/test/java/com/clickhouse/client/http/ClickHouseHttpClientTest.java [hunk 5]
```diff
@@ -143,65 +144,109 @@ public void testUserAgent() throws Exception {
...
-            String result = response.firstRecord().getValue(0).asString();
-            Assert.assertTrue(result.startsWith(client.getConfig().getProductName()));
-            Assert.assertTrue(result.indexOf("Http") > 0);
+
+        String selectQuery = "select http_user_agent from system.query_log where query='select ''" + uuid + "'''";
+        if (isCloud()) {
+            selectQuery = "SELECT http_user_agent FROM clusterAllReplicas(default, 'system', query_log) WHERE query_id='" + queryId + "'";
         }
 
+        int loopCount = 0;
```
Comment (mshustov) [inline clickhouse-http-client/src/test/java/com/clickhouse/client/http/ClickHouseHttpClientTest.java:162]: The same as above: avoid mixing the test suite and service code. We can create a helper that wraps test execution and retries until a successful assertion or a boundary condition.
Consider reusing this helper in all the tests in the file

File: clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIntegrationTest.java [hunk 6]
```diff
@@ -70,17 +69,17 @@ protected void checkRowCount(Statement stmt, String queryOrTableName, int expect
...
-        if (className.startsWith(CLASS_PREFIX)) {
-            className = className.substring(CLASS_PREFIX.length());
-        }
-        if (className.endsWith(CLASS_SUFFIX)) {
-            className = className.substring(0, className.length() - CLASS_SUFFIX.length());
-        }
-
-        this.dbName = "test_" + className.toLowerCase(Locale.ROOT);
-    }
+//    public JdbcIntegrationTest() {
```
Comment (mshustov) [inline clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIntegrationTest.java:72]: What is your plan for this commented-out snippet?

File: clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java (hunk 1 above)
Comment (mshustov) [reply clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/JdbcIssuesTest.java:136]: @Paultagoras Sergey's concern is still valid

**Review by mshustov — CHANGES_REQUESTED** [review]

File: clickhouse-client/src/test/java/com/clickhouse/client/ClientIntegrationTest.java (hunk 4 above)
Comment (mshustov) [reply clickhouse-client/src/test/java/com/clickhouse/client/ClientIntegrationTest.java:2679]: @Paultagoras I've found 106 `TODO` in the project. Without a link, it is hard to understand when and how these are going to be addressed (if ever)

File: .github/workflows/analysis.yml [hunk 7]
```diff
@@ -33,7 +33,7 @@ env:
 jobs:
   static:
     runs-on: ubuntu-latest
-    timeout-minutes: 15
+    timeout-minutes: 45
```
Comment (chernser) [inline .github/workflows/analysis.yml:36]: 😱

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -179,7 +179,7 @@ jobs:
       matrix:
         # most recent LTS releases as well as latest stable builds
         # https://github.com/ClickHouse/ClickHouse/pulls?q=is%3Aopen+is%3Apr+label%3Arelease
-        clickhouse: ["23.8", "24.3", "24.6", "latest"]
+        clickhouse: ["22.8", "23.3", "23.7", "latest"]
```
Comment (chernser) [inline .github/workflows/build.yml:182]: Please use existing one ` ["23.8", "24.3", "24.6", "latest"]`
I've updated it recently and your change is going to revert it =)

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -228,13 +231,74 @@ jobs:
...
+            **/target/failsafe-reports
+            **/target/surefire-reports
 
   test-jdbc-driver:
     runs-on: ubuntu-latest
     needs: compile
     strategy:
       matrix:
-        clickhouse: ["23.8", "24.3", "24.6", "latest"]
+        clickhouse: ["22.8", "23.3", "23.7", "latest", "cloud"]
```
Comment (chernser) [inline .github/workflows/build.yml:301]: please use `["23.8", "24.3", "24.6", "latest", "cloud"]`

File: clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseNodeTest.java [hunk 10]
```diff
@@ -393,15 +393,26 @@ public void testQueryWithSlash() throws URISyntaxException {
...
         ClickHouseVersion serverVersion = ClickHouseVersion
                 .of(System.getProperty("clickhouseVersion", "latest"));
         for (ClickHouseProtocol p : protocols) {
             if (p == ClickHouseProtocol.GRPC && !serverVersion.check("[21.1,)")) {
                 continue;
             }
-            ClickHouseNode node = getServer(ClickHouseProtocol.ANY, p.getDefaultPort());
+            int port = p.getDefaultPort();
+            if (isCloud()) {
+                port = 8433;
```
Comment (chernser) [inline clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseNodeTest.java:413]: It should be `8443`: 
```
    HTTP(8123, 8443, "http", "https"),
```

File: clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseServerForTest.java [hunk 11]
```diff
@@ -289,5 +351,45 @@ public static void afterSuite() {
...
+                    return true;
+                }
+            } catch (Exception e) {
+                LOGGER.warn("Failed to run query for testing...", e);
+            }
+
+            try {
+                Thread.sleep(15000);
+            } catch (InterruptedException e) {
+                LOGGER.error("Failed to sleep", e);
```
Comment (chernser) [inline clickhouse-client/src/test/java/com/clickhouse/client/ClickHouseServerForTest.java:389]: here we need exit immediately - interrupt is interrupt =)

File: clickhouse-http-client/src/test/java/com/clickhouse/client/http/ApacheHttpConnectionImplTest.java [hunk 12]
```diff
@@ -92,6 +88,7 @@ public void testConnection() throws Exception {
     }
 
     @Test(groups = { "integration" })
+    @Ignore("Need to disable the option to provide custom socket factory. note: need to remove it")
```
Comment (chernser) [inline clickhouse-http-client/src/test/java/com/clickhouse/client/http/ApacheHttpConnectionImplTest.java:91]: Why would we need to remove it? 
It is a bit strange to provide custom socket factory, but it was added for some reason. Certificates?

File: clickhouse-http-client/src/test/java/com/clickhouse/client/http/ClickHouseHttpClientTest.java [hunk 13]
```diff
@@ -365,16 +417,27 @@ public void testLogComment() throws ClickHouseException, IOException {
                     .query("SYSTEM FLUSH LOGS", uuid).executeAndWait()) {
             }
 
+            String selectQuery = "SELECT log_comment FROM system.query_log WHERE query_id = :qid";
+            if (isCloud()) {
+                selectQuery = "SELECT log_comment FROM clusterAllReplicas(default, 'system', query_log) WHERE log_comment != '' AND query_id = :qid";
+                try {
+                    Thread.sleep(30000);
+                } catch (InterruptedException e) {
+                    // ignore
```
Comment (chernser) [inline clickhouse-http-client/src/test/java/com/clickhouse/client/http/ClickHouseHttpClientTest.java:426]: exit here. interrupted exception may not be ignored.
